### ISSUE TITLE
False positive on `NoNameShadowing` with unused Unit as outer lambda parameter

### ISSUE THREAD

**Issue by Logima** [detekt/detekt#4692]
## Expected Behavior
Not to report NoNameShadowing, when outer lambda implicit parameter has type Unit and is unused.

## Observed Behavior
`test.kt:20:26: Name shadowed: implicit lambda parameter 'it' [NoNameShadowing]`

## Steps to Reproduce
Minimal (theoretical) example:
```kt
fun String.test(a: String.(Unit) -> Unit) {
    // ...
}
"".test { // it is Unit and unused
    chars().forEach { print(it) } // Name shadowed: implicit lambda parameter 'it' [NoNameShadowing]
}
```

## Context
The issue seems to be that Unit is explicitely defined. When lambda parameter type is empty, the rule works fine.
```kt
fun String.test(a: String.() -> Unit) {
    // ...
}
"".test { // it is Unit and unused
    chars().forEach { print(it) } // No error
}
```

## Your Environment
* Version of detekt used: `1.20.0-RC2`
* Version of Gradle used (if applicable): `7.4.2`

**Comment by kvn-stgl** [comment]
I'm not sure whether this is really a bug or not. If you explicitly define a parameter as Unit, there is an implicit lambda `it` with the value Unit and you are able to work with it. So there is a name shadowing.
 
In the other case, if you don't define a parameter, there will be no implicit lambda. So the following code should not work:
```kotlin
fun String.test(a: String.() -> Unit) {
    // ...
}
fun main() {
    "".test {
        println(it)
    }
}
```

**Comment by BraisGabin** [comment]
I agree with @kvn-stgl  here. Why is this a false positive? You are shadowing the it of the outer lambda so it should be flagged. If you don't want to shadow it add `_ - >`.

And why do you have unit as a parameter? Could you provide an use case?

**Comment by Logima** [comment]
> In the other case, if you don't define a parameter, there will be no implicit lambda. So the following code should not work:

Oh, I thought missing parameter and Unit meant practically the same thing, but apparently not.

> I agree with @kvn-stgl here. Why is this a false positive? You are shadowing the it of the outer lambda so it should be flagged. If you don't want to shadow it add `_ - >`.

That is an option. I understand now that technically the parameter is shadowed, but as the outer parameter is unused, I would prefer that it wouldn't result in `NoNameShadowing`.

> And why do you have unit as a parameter? Could you provide an use case?

The Unit parameter might be in an external library. Another use case is with Ktor routes, for example [get](https://github.com/ktorio/ktor/blob/main/ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/routing/RoutingBuilder.kt#L107). Simplified example from my application:

```kt
suspend fun wrapCall(call: ApplicationCall, callback: suspend (CallWrapper) -> Unit) {
    callback(CallWrapper(call))
}

routing {
    get("/path") {
        wrapCall(call) {
            processPath(it)
        }
    }
}
```

IntelliJ IDEA inspection NestedLambdaShadowedImplicitParameter doesn't flag this case. It shows warning only when the outer lambda parameter is used:
```kt
routing {
    get("/path") {
        println(it)
        wrapCall(call) {
            processPath(it) // "Implicit parameter 'it' of enclosing lambda is shadowed"
        }
    }
}
```

**Comment by BraisGabin** [comment]
I would like to know what other maintainers think about this but to me even if you are not using it you are shadowing that value and it's not that clear what `it` you are using.

**Comment by marschwar** [comment]
I agree with @BraisGabin . If `it` is of type unit I think it should not raise an issue, but in any other case it should.

**Comment by jruuttun** [comment]
Related: The [corresponding IDEA inspection was changed](https://youtrack.jetbrains.com/issue/KTIJ-11947/Should-not-report-implicit-it-is-shadowed-when-outer-it-is-not-used) to *not* report if outer 'it' is not used. I find this to be the better approach.

**Comment by masmil1988** [comment]
I too agree that this is a false positive. In facts this rule forces us to define a name when there's not an explicit clash among `it`s.
